saliency/base.py

```diff
@@ -25,18 +26,11 @@ def __init__(self, graph, session, y, x):
       graph: The TensorFlow graph to evaluate masks on.
       session: The current TensorFlow session.
       y: The output tensor to compute the SaliencyMask against. This tensor
-          should be of size 1.
+          should have the same first dimension (batch size) as x.
       x: The input tensor to compute the SaliencyMask against. The outer
           dimension should be the batch size.
     """
 
-    # y must be of size one, otherwise the gradient we get from tf.gradients
-    # will be summed over all ys.
-    size = 1
```

Comment:
can you write a unit test that covers the batch case as you expect? can you explain how you solve this in the description (without having to read stackoverflow)

Reply:
The existing integrated_gradients_test.py is now testing batching. I'm not batching across different images (there is still only one input image), but batching the x_steps (1000 in the test case) so the 1000 gradients are computed in one session run.

Do you mean I should add comments in the code to explain why batching the x_steps works or just in the description to this PR? If it's the latter, before we thought batching wouldn't work because if x has [batch_size, N] as its shape but y has shape [N] (this is the case because we are feeding y[0] to the constructor in the test), tf.gradient would sum over the first dimension of x. Now allow y to have shape [batch_size, N], we would get a list of gradients of size batch_size. This is what the stackoverflow link is saying too. TensorFlow has to support computing gradient in a batch.

Reply:
I also tried batching the nsamples when computing the smoothed mask, but that doesn't help much for smoothed integrated gradient because now we are computing nsamples * x_steps masks, which makes the input too big and slow to compute in one session run, so I added multithreading to do concurrent session run to compute smoothed masks.

Reply:
Ahhh got it, apologies for misunderstanding!

Reply:
No worries at all! I was also very confused. Mind approving the pool request?